pkg/processor/worker/worker.go

```diff
@@ -151,20 +151,25 @@ func (w *Worker) SupportsRestart() bool {
 	return w.runtime.SupportsRestart()
 }
 
-func (w *Worker) Drain() error {
-	err := w.runtime.Drain()
-	if err == nil {
+func (w *Worker) IsDrained() bool {
+	w.drainedLock.Lock()
+	defer w.drainedLock.Unlock()
 
-		w.setDrained(true)
-	}
-	return err
+	return w.isDrained
 }
 
-func (w *Worker) IsDrained() bool {
+func (w *Worker) Drain() error {
 	w.drainedLock.Lock()
 	defer w.drainedLock.Unlock()
 
-	return w.isDrained
+	if !w.isDrained {
+		err := w.runtime.Drain()
+		if err == nil {
+			w.isDrained = true
```

Comment:
Use `SetDrained(true)` that already uses the locks.
Then the locks in lines 162 won't be needed.

Reply:
same here